internal/pkg/coordinator/monitor.go

```diff
@@ -0,0 +1,369 @@
+					pt.canceller = nil
+					pt.cord = nil
+				}
+				res <- pt
+				return
+			}
+			if pt.cord == nil {
+				cord, err := m.factory(p)
+				if err != nil {
+					l.Err(err).Msg("failed to start coordinator")
+					_ = dl.ReleasePolicyLeadership(ctx, m.bulker, pt.id, m.agentMetadata.Id, m.leaderInterval, dl.WithIndexName(m.leadersIndex))
+					res <- pt
+					return
+				}
+
+				cordCtx, canceller := context.WithCancel(ctx)
+				go runCoordinator(cordCtx, cord, l, m.coordRestartDelay)
+				go runCoordinatorOutput(cordCtx, cord, m.bulker, l, m.policiesIndex)
+				pt.cord = cord
+				pt.canceller = canceller
+			}
+			res <- pt
+		}(p, pt)
+	}
+	for range lead {
+		r := <-res
+		if r.cord == nil {
+			// either failed to take leadership or lost leadership
+			delete(m.policies, r.id)
+		} else {
+			m.policies[r.id] = r
+		}
+	}
+	return nil
+}
+
+// releaseLeadership releases current leadership
+func (m *monitorT) releaseLeadership() {
+	var wg sync.WaitGroup
+	wg.Add(len(m.policies))
+	for _, pt := range m.policies {
+		go func(pt policyT) {
+			if pt.cord != nil {
+				pt.canceller()
+			}
+			// uses a background context, because the context for the
+			// monitor will be cancelled at this point in the code
+			err := dl.ReleasePolicyLeadership(context.Background(), m.bulker, pt.id, m.agentMetadata.Id, m.leaderInterval, dl.WithIndexName(m.leadersIndex))
+			if err != nil {
+				l := m.log.With().Str(dl.FieldPolicyId, pt.id).Logger()
+				l.Err(err).Msg("failed to release leadership")
+			}
+			wg.Done()
+		}(pt)
+	}
+	wg.Wait()
+}
+
+func (m *monitorT) calcMetadata() {
+	m.agentMetadata = model.AgentMetadata{
+		Id:      m.fleet.Agent.ID,
+		Version: m.fleet.Agent.Version,
+	}
+	hostname := m.fleet.Host.Name
+	if hostname == "" {
+		h, err := os.Hostname()
+		if err != nil {
+			m.log.Err(err).Msg("failed to get hostname")
+		}
+		hostname = h
+	}
+	ips, err := m.getIPs()
+	if err != nil {
+		m.log.Err(err).Msg("failed to get ip addresses")
+	}
+	m.hostMetadata = model.HostMetadata{
+		Id:           m.fleet.Host.ID,
+		Name:         hostname,
+		Architecture: runtime.GOOS,
+		Ip:           ips,
+	}
+}
+
+func (m *monitorT) getIPs() ([]string, error) {
+	ifaces, err := net.Interfaces()
+	if err != nil {
+		return nil, err
+	}
+	ips := []string{}
+	for _, i := range ifaces {
+		addrs, err := i.Addrs()
+		if err != nil {
+			continue
+		}
+		for _, addr := range addrs {
+			var ip net.IP
+			switch v := addr.(type) {
+			case *net.IPNet:
+				ip = v.IP
+			case *net.IPAddr:
+				ip = v.IP
+			}
+			if ip != nil {
+				ips = append(ips, ip.String())
+			}
+		}
+	}
+	return ips, nil
+}
+
+func runCoordinator(ctx context.Context, cord Coordinator, l zerolog.Logger, d time.Duration) {
+	for {
+		l.Info().Str("coordinator", cord.Name()).Msg("starting coordinator for policy")
+		err := cord.Run(ctx)
+		if err != context.Canceled {
+			l.Err(err).Msg("coordinator failed")
+			select {
+			case <-time.After(d):
+				break
+			case <-ctx.Done():
```

Comment:
maybe use timer that can reset here, otherwise, time.After will keep ticking until fired